### PR TITLE
Fogl 1641 : The destinations table is obsolete and will be removed

### PR DESCRIPTION
**Pull request by sstefano69**
**ready**

### CODE REVIEW

**Comment by praveen-garg** [comment]
We must to thorough testing of it, to avoid any regression.

**Comment by sstefano69** [comment]
Hi @praveen-garg : 
indeed I tested :
tests :
* pi_server - readings
vpi_server - statistics
* ocs
* c++ - readings
* upgrade/downgrade
** vbox - postgress
** raspbian - sqllite

* System tests

**Comment by praveen-garg** [comment]
@sstefano69  I think https://github.com/foglamp/FogLAMP/pull/1050 is also doing work on streams (FOGL-1748) for postgres,  could you please check and see if we can close that PR in favor of this?

File: scripts/plugins/storage/sqlite/upgrade/12.sql [hunk 1]
```diff
@@ -0,0 +1,36 @@
+DROP TABLE IF EXISTS foglamp.destinations;
+DROP INDEX IF EXISTS foglamp.fki_streams_fk1;
```
Comment (praveen-garg) [inline scripts/plugins/storage/sqlite/upgrade/12.sql]: did we check this? I think its not complaining as we have `IF EXISTS`  you may need to remove `foglamp.`

File: scripts/plugins/storage/sqlite/downgrade/11.sql [hunk 2]
```diff
@@ -0,0 +1,53 @@
+CREATE TABLE foglamp.destinations (
+    id            INTEGER                     PRIMARY KEY AUTOINCREMENT,                  -- Sequence ID
+    type          smallint                    NOT NULL DEFAULT 1,                         -- Enum : 1: OMF, 2: Elasticsearch
+    description   character varying(255)      NOT NULL DEFAULT '',                        -- A brief description of the destination entry
+    properties    JSON                        NOT NULL DEFAULT '{ "streaming" : "all" }', -- A generic set of properties
+    active_window JSON                        NOT NULL DEFAULT '[ "always" ]',            -- The window of operations
+    active        boolean                     NOT NULL DEFAULT 't',                       -- When false, all streams to this destination stop and are inactive
+    ts            DATETIME DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'localtime')));         -- Creation or last update
+
+INSERT INTO foglamp.destinations ( id, description )
```
Comment (praveen-garg) [inline scripts/plugins/storage/sqlite/downgrade/11.sql]: why do we need this?

File: scripts/plugins/storage/sqlite/downgrade/11.sql [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    ts             DATETIME DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'localtime')), -- Creation or last update
+    CONSTRAINT streams_fk1 FOREIGN KEY (destination_id)
+    REFERENCES destinations (id) MATCH SIMPLE
+            ON UPDATE NO ACTION
+            ON DELETE NO ACTION );
+
+INSERT INTO foglamp.streams
+        SELECT
+            id,
+            0,
```
Comment (praveen-garg) [inline scripts/plugins/storage/sqlite/downgrade/11.sql]: why hard-coded 0

File: scripts/plugins/storage/postgres/downgrade/11.sql [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+       CONSTRAINT strerams_pkey PRIMARY KEY (id),
+       CONSTRAINT streams_fk1 FOREIGN KEY (destination_id)
+       REFERENCES foglamp.destinations (id) MATCH SIMPLE
+               ON UPDATE NO ACTION
+               ON DELETE NO ACTION );
+
+INSERT INTO foglamp.streams
+        SELECT
+            id,
+            0,
```
Comment (praveen-garg) [inline scripts/plugins/storage/postgres/downgrade/11.sql]: why hard-coded 0?

File: scripts/plugins/storage/postgres/downgrade/11.sql [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+       id            integer                     NOT NULL DEFAULT nextval('foglamp.destinations_id_seq'::regclass),   -- Sequence ID
+       type          smallint                    NOT NULL DEFAULT 1,                                                  -- Enum : 1: OMF, 2: Elasticsearch
+       description   character varying(255)      NOT NULL DEFAULT ''::character varying COLLATE pg_catalog."default", -- A brief description of the destination entry
+       properties    jsonb                       NOT NULL DEFAULT '{ "streaming" : "all" }'::jsonb,                   -- A generic set of properties
+       active_window jsonb                       NOT NULL DEFAULT '[ "always" ]'::jsonb,                              -- The window of operations
+       active        boolean                     NOT NULL DEFAULT true,                                               -- When false, all streams to this destination stop and are inactive
+       ts            timestamp(6) with time zone NOT NULL DEFAULT now(),                                              -- Creation or last update
+       CONSTRAINT destination_pkey PRIMARY KEY (id) );
+
+INSERT INTO foglamp.destinations ( id, description )
```
Comment (praveen-garg) [inline scripts/plugins/storage/postgres/downgrade/11.sql]: why this INSERT is needed?

File: scripts/plugins/storage/sqlite/upgrade/12.sql (hunk 1 above)
Comment (sstefano69) [reply scripts/plugins/storage/sqlite/upgrade/12.sql]: I tested, I retested right now to be sure :  

Downgrade :
``` 
Detected 'sqlite' FogLAMP DB schema change from version [12] to [11], applying Upgrade/Downgrade ...
FogLAMP DB schema has been downgraded to version [11]
Starting FogLAMP v1.3.1......
FogLAMP started.
```

Upgrade : 
```
foglamp@ubuntu:~$ ${FOGLAMP_ROOT}/scripts/foglamp start
Detected 'sqlite' FogLAMP DB schema change from version [11] to [12], applying Upgrade/Downgrade ...
FogLAMP DB schema has been upgraded to version [12]
Starting FogLAMP v1.3.1......
FogLAMP started.
foglamp@ubuntu:~$
```

the table was removed :
```
foglamp@ubuntu:~$ export FOGLAMP_ROOT=/home/foglamp/Development/FogLAMP;\
> cmd_sql='SELECT *  FROM destinations;';\
> sqlite3 -column  -header   -cmd ".timeout 5000"  "file:${FOGLAMP_ROOT}/data/foglamp.db?mode=ro&nolock=1"  "${cmd_sql}"
Error: no such table: destinations
foglamp@ubuntu:~$
```

File: scripts/plugins/storage/sqlite/downgrade/11.sql (hunk 2 above)
Comment (sstefano69) [reply scripts/plugins/storage/sqlite/downgrade/11.sql]: the foglamp.streams has a FK constrains when we downgrade :
```
...
    CONSTRAINT streams_fk1 FOREIGN KEY (destination_id)
    REFERENCES destinations (id) MATCH SIMPLE
...
```
so we need to create a dummy row in the destinations table to handle the reference for the rows that are already in the streams before the downgrade.

File: scripts/plugins/storage/sqlite/downgrade/11.sql (hunk 3 above)
Comment (sstefano69) [reply scripts/plugins/storage/sqlite/downgrade/11.sql]: explained in the previous.

File: scripts/plugins/storage/postgres/downgrade/11.sql (hunk 4 above)
Comment (sstefano69) [reply scripts/plugins/storage/postgres/downgrade/11.sql]: explained in the previous.

File: scripts/plugins/storage/postgres/downgrade/11.sql (hunk 5 above)
Comment (sstefano69) [reply scripts/plugins/storage/postgres/downgrade/11.sql]: explained in the previous.

File: scripts/plugins/storage/sqlite/upgrade/12.sql (hunk 1 above)
Comment (praveen-garg) [reply scripts/plugins/storage/sqlite/upgrade/12.sql]: My comment was for ` DROP INDEX IF EXISTS foglamp.fki_streams_fk1; `

**Pull request merged**